src/glpk/env/alloc.c

```diff
@@ -45,74 +46,141 @@
 *  for book-keeping the memory usage statistics. */
 
 #ifdef HAVE_ENV
+
+/**
+ * "Unwrap" the memory block descriptor
+ */
+static inline MBD* ptr_to_mbd(void* ptr)
+{
+    MBD* mbd = (MBD *)((char *)ptr - MBD_SIZE);
+    xassert(mbd->self == mbd);
+    return mbd;
+}
+
+/**
+ * Take the mbd off the environment. Must only ever try to remove an mdb from the environment
+ * it was allocated on
+ */
+static inline void _remove_from_env(MBD* mbd, ENV* env)
+{
+    xassert(mbd->env == env);
+
+    MBD* before = mbd->prev;
+    MBD* after = mbd->next;
+
+    /* unlink ourselves */
+    if(after) {
+        after->prev = before;
+    }
+    if(before) {
+        before->next = after;
+    } else {
+        env->mem_ptr = after;
+    }
+
+    /* update counters */
+    if (get_mem_count() < 1 || get_mem_total() < mbd->size) {
+        xerror("unlinking mbd: memory allocation error\n");
+    }
+    add_mem_count(-1);
+    add_mem_total(-(mbd->size));
+}
+
+/**
+ * Free an mbd from an env. If the mbd's env and the current env mismatch, the block will *not* be deallocated yet,
+ * instead the deallocation will be deferred until the environment is freed
+ */
+static inline void* _free_mbd(MBD* mbd, ENV* env) {
+    if(mbd->env != env) {
+        return NULL;
+    }
+    _remove_from_env(mbd, env);
+    free(mbd);
+    return NULL;
+}
+
+static inline void _check_allocation(ENV* env, size_t size) 
+{
+    if (size > SIZE_T_MAX - MBD_SIZE)
+        xerror("block too large\n");
+    if (size + MBD_SIZE > get_mem_limit() - get_mem_total())
+        xerror("memory allocation limit exceeded\n");
+    if (get_mem_count() == INT_MAX)
+        xerror("too many memory blocks allocated\n");
+}
+
+static inline void _prepend_mbd_to_env(MBD* mbd, ENV* env, size_t size)
+{
+    xassert(mbd != NULL && env != NULL);
```

Comment:
Splitting this into two separate asserts can help in debugging. If the conditions are combined in a single assert, you will not know which one of the conditions failed.

Reply:
fixed in 9281f64